controllers/configuration/backend/backend.go

```diff
@@ -0,0 +1,206 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package backend
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+	"strings"
+
+	"github.com/hashicorp/hcl/v2"
+	"github.com/hashicorp/hcl/v2/gohcl"
+	"github.com/hashicorp/hcl/v2/hclparse"
+	"github.com/hashicorp/hcl/v2/hclwrite"
+	"github.com/oam-dev/terraform-controller/api/v1beta2"
+	"github.com/pkg/errors"
+	"github.com/zclconf/go-cty/cty"
+	"github.com/zclconf/go-cty/cty/gocty"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+var backendInitFuncMap = map[v1beta2.BackendType]*backendInitFunc{
+	"kubernetes": {
+		InitFuncFromHCL:  newK8SBackendFromHCL,
+		InitFuncFromConf: newK8SBackendFromConf,
+	},
+}
+
+type Backend interface {
+	HCL() string
+	GetTFStateJSON(ctx context.Context) ([]byte, error)
+	CleanUp(ctx context.Context) error
+}
+
+type backendInitFunc struct {
+	InitFuncFromHCL  func(string, *ParsedBackendConfig, client.Client) (Backend, error)
+	InitFuncFromConf func(string, interface{}, client.Client) (Backend, error)
+}
+
+type ParsedBackendConfig struct {
+	Name   string   `hcl:"name,label"`
+	Remain hcl.Body `hcl:",remain"`
+}
+
+func (conf ParsedBackendConfig) getAttrValue(key string) (*cty.Value, error) {
+	attrs, diags := conf.Remain.JustAttributes()
+	if diags.HasErrors() {
+		return nil, diags
+	}
+	attr := attrs[key]
+	if attr == nil {
+		return nil, fmt.Errorf("cannot find attr %s", key)
+	}
+	v, diags := attr.Expr.Value(nil)
+	if diags.HasErrors() {
+		return nil, diags
+	}
+	return &v, nil
+}
+
+func (conf ParsedBackendConfig) getAttrString(key string) (string, error) {
+	v, err := conf.getAttrValue(key)
+	if err != nil {
+		return "", err
+	}
+	result := ""
+	err = gocty.FromCtyValue(*v, &result)
+	return result, err
+}
+
+// ParseConfigurationBackend parses backend Conf from the v1beta2.Configuration
+func ParseConfigurationBackend(configuration *v1beta2.Configuration, k8sClient client.Client) (Backend, error) {
+	backend := configuration.Spec.Backend
+
+	switch {
+
+	case backend == nil || (backend.Inline == "" && backend.BackendType == ""):
+		// use the default k8s backend
+		if configuration.Spec.Backend != nil {
+			if configuration.Spec.Backend.SecretSuffix == "" {
+				configuration.Spec.Backend.SecretSuffix = configuration.Name
+			}
+			configuration.Spec.Backend.InClusterConfig = true
+		} else {
+			configuration.Spec.Backend = &v1beta2.Backend{
+				SecretSuffix:    configuration.Name,
+				InClusterConfig: true,
+			}
+		}
+		return newDefaultK8SBackend(configuration.Spec.Backend.SecretSuffix, k8sClient), nil
+
+	case backend.Inline != "" && backend.BackendType != "":
```

Comment:
We don't need `backend.Inline != ""` in the following cases.

Reply:
Now we have for cases:

```go
	case backend == nil || (backend.Inline == "" && backend.BackendType == ""):

	case backend.Inline != "" && backend.BackendType != "":

	case backend.Inline != "":

	case backend.BackendType != "":
```

Do you mean that we should change them to the following?

```go
	case backend == nil || (backend.Inline == "" && backend.BackendType == ""):

	case backend.Inline != "" && backend.BackendType != "":

	case backend.BackendType != "":
        
        default:

```

Reply:
No, you are right.